base/compiler/ssair/passes.jl

```diff
@@ -1648,18 +1647,23 @@ function try_resolve_finalizer!(ir::IRCode, alloc_idx::Int, finalizer_idx::Int,
             if bb == finalizer_bb
                 s = finalizer_idx
             end
-            check_range_nothrow(s, e) || return nothing
+            all(s:e) do sidx::Int
+                sidx == finalizer_idx && return true
+                sidx == alloc_idx && return true
+                return is_nothrow(ir, SSAValue(sidx))
+            end && continue
+
+            # An exception may be thrown between the finalizer registration and the point
+            # where the object’s lifetime ends (`insert_idx`): In such cases, we can’t
+            # remove the finalizer registration, but we can still insert a `Core.finalize`
+            # call at `insert_idx` while leaving the registration intact.
+            newinst = add_flag(NewInstruction(Expr(:call, GlobalRef(Core, :finalize), finalizer_stmt.args[3]), Nothing), flag)
```

Comment:
This can have a surprising performance cost, since running finalizers are normally O(1), but this call is O(n). We would need to reimplement the way finalizers are registered.

To make this particular case legal, we also need to prove that we are only going to run the expected finalizer that it originally detected being registered (no other finalizers can be registered along side it, though those may be considered lifetime escapes already?) as other finalizers might not be allowed to run on this Task. There is also an issue that inference did the analysis in the tls world, but this runs them in the unknown future world.

Reply:
> To make this particular case legal, we also need to prove that we are only going to run the expected finalizer that it originally detected being registered (no other finalizers can be registered along side it, though those may be considered lifetime escapes already?

Regarding this, the escape check in `sroa_mutables!` ensures that the object in question only has a single finalizer.

> There is also an issue that inference did the analysis in the tls world, but this runs them in the unknown future world.

The docstring for `finalizer` states:

> > Note that there is no guaranteed world age for the execution of `f`. It may be called in the world age in which the finalizer was registered or any later world age.

So, it should be acceptable for the world in which the finalizer is inferred and the world in which the finalizer is executed to be different?

Reply:
> Regarding this, the escape check in sroa_mutables! ensures that the object in question only has a single finalizer.

Okay good. I wasn't sure if the analysis was total (guaranteed that there was no other finalizer), or simply that it only uses one of the finalizers, depending on what it can detect locally

> So, it should be acceptable for the world in which the finalizer is inferred and the world in which the finalizer is executed to be different?

It is acceptable to be different, but therefore not acceptable for inference to make decisions based on its guaranteed behavior, and then to not enforce the world of the runtime. We could probably change `finalize`, but the main concern is that `finalize` is already slow and undesirable compared to letting the GC do its work to sweep them as a batch, and we don't necessarily want to make `finalizer` slower just to support that function. So we can try to forward the function in `finalizer` directly here, but we need some (fast) way to also undo that finalizer registration (or delay it until either an error happens or some other branch which won't execute this finalize call insertion location)

Reply:
Yeah, it looks like `finalize` is pretty slow. I tried to optimize it to some extent in 754a8bdd613cad562312e222a121c43f542bcec3, but it’s still about 2.5 times slower compared to `master` in the following benchmark:
```julia
​mutable struct AtomicCounter
    @atomic count::Int
end

const counter = AtomicCounter(0)

function withfinalizer(x)
    xs = finalizer(Ref(x)) do obj
        Base.@assume_effects :nothrow :notaskstate
        @atomic counter.count += obj[]
    end
    println(devnull, xs[])
    return xs[]
end

@benchmark withfinalizer(1)
```
(even though I confirmed that `withfinalizer(1)` certainly hits the fast path added in the commit)

So, as you mentioned, we’ll probably need to implement a way to cancel the finalizer registration or delay the registration itself. The former sounds a bit simpler, so I’ll explore that approach.

Reply:
I tried out the approach of canceling the finalizer registration (0f1da0e3d7f567a8b0f18c517cd7a0de87c8e61c). What do you think?